syntax = "proto3";

package presence;
import "google/protobuf/timestamp.proto";

enum ConnectionStatus{
  UNDETERMINED = 0;
  CONNECTED = 1;
  AWAY = 2;
  DISCONNECTED = 3;
}

message PresenceModel{
  string user_id = 1;
  string device_id = 2;
  ConnectionStatus status = 3;
  int64 last_seen_timestamp = 4; 
}

message UpdateUserConnectionRequest{
  string device_id = 1;
  //user_id should normally come from the authcontext
  string user_id = 2;
  ConnectionStatus status = 3;
  int64 last_seen_timestamp = 4;
}

message UpdateUserConnectionReply{
  ConnectionStatus status = 1;
}

message RetrieveUserPresenceByIdRequest{
  string user_id = 1;
}

message RetrieveUserPresenceByIdResponse{
  PresenceModel model = 1;
}

service PresenceRecorder {
  rpc UpdateStatus(UpdateUserConnectionRequest) returns (UpdateUserConnectionReply) {}
}

service PresenceRetriever{
  rpc RetrieveUserPresenceByUserId(RetrieveUserPresenceByIdRequest) returns (RetrieveUserPresenceByIdResponse){}
}